### PR TITLE
Docs: Contribution Guide

### PR DESCRIPTION
**Pull request by webecke**
## Overview

This PR introduces a Contribution Guide that makes official a lot of assumed practices on the Autograder development team, and helps new developers know how to contribute.

**Please give your feedback! I don't want this to be just the _Dallin said so_ document. What needs changing? What's redundant or too much? What's missing?**

## Details

- Introduces a PR template (which in theory will work once this is merged)
- Contribution Guide gives pointers to new developers
- Introduces commit message guidelines
- Formalizes the PR process
- Formalizes the issue-name-prefix practice currently in place
- Explains how to add a person to the AboutPage
- Says tests are required for new features
- A bunch more, you should just go read the new `CONTRIBUTING.md`

## Future Work

Two things come to mind as I've been working on this, which should be separate issues
- #461 
- Testing Guide, to help developers know how to get started testing, and provide consistent testing approach

_**Also, this branch was based on the `README-improvements` branch, not main, so a rebase might be required if this PR is finished before the other one**_

### CODE REVIEW

File: docs/CONTRIBUTING.md
Comment (19mdavenport) [inline docs/CONTRIBUTING.md:101]: I think it would be useful to allow the naming of branches by GitHub--The `Create a branch` button on the issue page

File: docs/CONTRIBUTING.md
Comment (19mdavenport) [inline docs/CONTRIBUTING.md:109]: (Half sarcastically) Might there be situations where this wouldn't apply? Might this be better worded with _should_ rather than _must_?

File: docs/CONTRIBUTING.md
Comment (19mdavenport) [inline docs/CONTRIBUTING.md:156]: I think I would rather see a slightly more descriptive pull request name (could be the same name as the issue, especially if the name of the branch is the same as the name of the issue)

File: docs/CONTRIBUTING.md
Comment (19mdavenport) [inline docs/CONTRIBUTING.md:166]: This may change after I leave, but at least as far as these things are concerned, I would like to be requested on all PRs to the main branch. This could be worded as "The auto-grader administrator TA" or similar.

Comment (19mdavenport) [inline docs/CONTRIBUTING.md:166]: Maybe this is me being tyrannical (and please tell me if you disagree), but I would like to wait for a review from the admin before merging (specifically if the pull request is to main--if its to a different branch then just the owner of the other branch would be necessary). Technically GitHub doesn't require that, but it might be good.

Comment (ThanGerlek) [reply docs/CONTRIBUTING.md:166]: Actually I think GitHub can enforce that, if we make you/the lead TA/whoever the `CODEOWNER` of everything in the repo. According to GPT there's a **Require review from Code Owners** option. And no, I don't think that's tyrannical. If anything, my concern is that it might become a bottleneck, but as long as the admin is on top of things, I think it's a good idea.

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [inline docs/CONTRIBUTING.md:161]: lol nother typo lolololol

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [inline docs/CONTRIBUTING.md:169]: Are we using joules or foot-pounds?

(Or ergs, if we're feeling really obscure)

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [inline docs/CONTRIBUTING.md:192]: Merge this line with the previous one.

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [inline docs/CONTRIBUTING.md:204]: Then re-request a review.

**Review by ThanGerlek — CHANGES_REQUESTED** [review]

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:101]: I agree that it's good to use issue numbers, but we also occasionally create PRs to address topics without creating an "Issue." I don't see a purpose in _requiring_ that an issue be created for every branch. The PR that will be generated from a branch is provides enough context. 

If you need help tracking stale branches, try the following resources:
* [GitHub Branches Page](https://github.com/softwareconstruction240/autograder/branches) lists all branches, and identifies them as "stale" or "active"
* `git log --oneline --graph --all --first-parent` can be used to visualize the relationships between many branches simultaneously. Scroll down the first page or so to start seeing the staler branches.
* We've enabled the feature to "automatically delete a branch when its PR is merged." This keeps the total number of branches down to _only_ those which are being worked on and those that never got finished.

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:109]: I agree with @19mdavenport. In general, I don't like saying absolute statements and telling people what to do. I prefer to recommend good approaches and recognize that all kinds of unique cases come up. I prefer to assume that the people working on this project are smart, ambitious, capable TAs who want to make a difference and learn stuff along the way.

Additionally, I don't feel like there are enough untested portions of our current code base that we can't exactly demonstrate this principle by example. I absolutely agree on the vision of having strong testing coverage over our codebase.

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:156]: Agreed.

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:166]: Should we attempt to codify this structure by adjusting the "required reviews" rules on the branch protections?

Comment (frozenfrank) [reply docs/CONTRIBUTING.md:166]: I love and support this discussion!

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:34]: Lol. This is an open PR. Fixing typos seems like a great thing to "just push" to the branch and make it go away.

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:62]: I like this kind of thinking 💡

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:119]: This is a contested topic, and I'm not sure any "right" answer exists. Sometimes, there are reasons to include comments, and sometimes that's not the best way to move forward. However, as you tried to argue recently in softwareconstruction240/checkstyle#3, sometimes the "most correct" way is not the way to move forward in the present moment.

The following summary of the "code comments are bad" debate highlights that there is a debate. Trying to decide one "correct" answer in a contested arena is going to be difficult.

## Rundown of Key Arguments from Both Sides of the Debate
> Credit: This section was kindly produced by [ChatGPT](https://chatgpt.com/share/672c723a-b5bc-800c-9af4-ee7dd9302e61).

Here’s a rundown of arguments from both sides of the debate on whether inline comments are a code smell or a readability enhancer:

### Arguments that Inline Comments are a Code Smell

1. **Code Should Be Self-Explanatory**  
   Advocates of this view believe that code itself should be clean and readable enough that it doesn't require inline comments. If comments are necessary to explain code, it may indicate the code is too complex or poorly structured.

2. **Comments Can Become Outdated**  
   Code is often updated, but comments may not be maintained with the same rigor, leading to situations where comments misrepresent the code. This inconsistency can confuse developers, especially in collaborative environments.

3. **Comments Might Encourage Bad Practices**  
   Relying on comments to explain complex logic can discourage developers from refactoring the code to make it simpler. Inline comments can sometimes mask bad design choices rather than addressing the underlying issues.

4. **Brevity and Focus**  
   Reducing the number of comments can keep code more focused and to the point, avoiding distractions. Some believe that excessive comments make it harder to quickly understand the main structure and purpose of the code.

### Arguments that Inline Comments Enhance Readability

1. **Contextual Clarification**  
   Inline comments can provide context or explain why certain decisions were made, especially when dealing with intricate or non-obvious logic, external dependencies, or unusual design choices that are not immediately clear.

2. **Onboarding and Collaboration**  
   For new team members or contributors unfamiliar with a codebase, inline comments can significantly reduce the learning curve by explaining complex sections. This can be particularly helpful in large teams or open-source projects.

3. **Code Can’t Capture Intent Fully**  
   Even well-written code may not always convey the intent behind it. Inline comments can clarify the purpose behind a line or block, helping future developers understand the reasoning rather than just the mechanics.

4. **Highlighting Workarounds or Known Issues**  
   Sometimes, developers implement workarounds for known issues or bugs. Inline comments can explain these so future developers aren’t left puzzled by non-standard implementations or don’t accidentally remove necessary code quirks.

Ultimately, whether inline comments are considered a code smell often depends on the codebase, team culture, and specific project requirements.

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:120]: It is a good policy to push all code at the end of a coding session, and is consistent with BYU's intellectual property policy to which we have all already agreed.

> ### Ownership
> Pursuant to law and university policy, **any work** (whether a Technical Work or a Creative Work) prepared by University Personnel within the scope of their employment, without an express agreement specifying otherwise, **is work for hire owned by the university**. 
> — [BYU Intellectual Property Policy](https://policy.byu.edu/view/intellectual-property-policy) (emphasis added)

Pushing the code into the BYU managed repo is one way we can give that IP to BYU. Conversely, keeping the work local is one way of keeping the IP from being in BYU's ownership.

Note that this policy doesn't specify that we have to push after "every commit," but only at the end of a coding session. This policy also doesn't restrict rebases. Rebasing a non-collaborative branch of any kind (feature or otherwise), requires a `git push --force`, but is otherwise safe the confident `git` user. This policy also doesn't preclude pushing the commits to other branches representing work-in-progress (WIP; this is another kind of branch besides feature branches).

This policy just requires that "the code is pushed," and that requirement is very consistent with the nature of our work in a managed environment for an employer.

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:124]: Good point, @ThanGerlek. I had to ask to discover the difference, and I was able to verify that "imperative statement" is the standard.

## In the context of authoring git commit messages, what's the difference between using "a present tense verb" and an "imperative statement?"

> Credit: The following research was kindly generated by [ChatGPT](https://chatgpt.com/share/672c7559-df04-800c-8ea3-60a17daad887).

In git commit messages, using a *present tense verb* and an *imperative statement* often overlap, but they serve slightly different purposes:

1. **Present Tense Verb**: Using the present tense describes what the commit does. For example, saying "Fixes bug in login flow" states in a straightforward way what the commit achieves.

2. **Imperative Statement**: Using the imperative mood (e.g., "Fix bug in login flow") gives a command or instruction, as if you’re telling Git (or a reader) what to do. This is preferred in git commit messages because it reads like an instruction—"If applied, this commit will..."—and maintains consistency with Git’s own conventions and tooling.

The imperative mood usually implies present tense, so in practice, it’s both present and imperative, making it a natural choice for clarity.

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:143]: This actually isn't an issue.

The GitHub markdown spec for [block quotes](https://github.github.com/gfm/#block-quote-marker) specifically describe a [**Laziness Clause**](https://github.github.com/gfm/#example-210). This means that consecutive lines on a block quote do not need to continue repeating the `>` character.

The result renders properly:
image

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:123]: I think the discussion in this section is just fine and appropriate for the level of engagement we're expecting on this project. We don't need anyone police-ing people on following an exact convention.

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:192]: Conversely, split the previous line into two separate lines. Emphasize that changed systems should be explicitly verified in addition to generally testing that "nothing broke."

File: docs/CONTRIBUTING.md
Comment (frozenfrank) [reply docs/CONTRIBUTING.md:123]: The advice provided by the Git User Manual is sound. It specifically— and importantly— mentions having a blank line between the summary of the commit and any additional details provided on additional lines.

> ### Creating good commit messages
> Though not required, it’s a good idea to begin the commit message with a single short (no more than 50 characters) line summarizing the change, followed by a blank line and then a more thorough description. The text up to the first blank line in a commit message is treated as the commit title, and that title is used throughout Git.
> — [Git User Manual](https://git-scm.com/docs/user-manual#creating-good-commit-messages)

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:119]: I'm not the biggest fan of inline comments, and the principle should be self documenting code. But I think there's definitely some places where it's not realistic. The way this section was worded was my attempt to encourage self documenting code, but require comments where it's not possible or difficult (In particular, I'm thinking of CSS, but this could happen other places). I'll go ahead and try to rework this until an official decision is made.

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:120]: The IP Policy is a great reminder, I'm going to add that to the bottom of the document.

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:109]: I'll go ahead and change it to should. I was just following the list of things @frozenfrank put on the issue for this PR.

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:34]: fixed

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:49]: This was supposed to be a broad overview of the process, which isn't really going to be known until after our meeting today.

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:68]: fixed

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:147]: fixed

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:120]: I split the line, attempting to be more explicit in what we're expecting.

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:124]: fixed

Comment (webecke) [reply docs/CONTRIBUTING.md:124]: I knew the way I had it written was weird, thanks for pointing this out. 

fixed

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:167]: I'm not an English major, what's the typo here?

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md]: I split the line

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [reply docs/CONTRIBUTING.md:167]: "Submittted" has three T's. That's it. Not exactly a game changer.

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [reply docs/CONTRIBUTING.md:175]: Nvm, that was a terrible joke

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [reply docs/CONTRIBUTING.md:91]: Fair enough, I just thought it would make sense to introduce the terminology. This was also related to something else I wanted to mention in the meeting, but it became moot partway through our discussion, and so no longer matters.

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [reply docs/CONTRIBUTING.md:123]: I get that there's an argument, I just don't think the CONTRIBUTING.md or this PR is the place. Hence the suggestion to link to the style guide.

I don't feel strongly here, especially since we decided to prioritize simplicity and brevity for new devs, so I'd be fine with a slight rewording or even just leaving it as it is.

File: docs/CONTRIBUTING.md
Comment (webecke) [reply docs/CONTRIBUTING.md:167]: Oh lol, I see it now 😂

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [reply docs/CONTRIBUTING.md:127]: Agreed, especially with focusing on simplicity for new devs.

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [reply docs/CONTRIBUTING.md:172]: NOTE: as of the dev meeting, this is outdated.

Comment (ThanGerlek) [reply docs/CONTRIBUTING.md:172]: NOTE: as of the dev meeting, this is outdated.

**Comment by frozenfrank** [comment]
This PR has been updated to merge into the #452 (which it was based on) so that the PR diff views appear properly.

File: docs/CONTRIBUTING.md
Comment (ThanGerlek) [inline docs/CONTRIBUTING.md:4]: ❤️

**Review by ThanGerlek — APPROVED** [review]

**Comment by webecke** [comment]
I can't figure out whats going on with the diff on this PR. it should only have a few markdown files different from `451-improve-readmes`, especially since I just merged main on both. But like half the files have big differences on this branch. Anyone have a clue why?

**Review by frozenfrank — APPROVED** [review]
Thank you for putting this together!

I made the last adjustments I felt necessary. At this point, are there any other blocking requirements to merge this PR?

**Pull request merged**
